File: c/bits.cpp
#include "stdafx.h"
#include "bits.h"
#include <bitset>


namespace bits
{
	std::uint64_t toFlag(std::uint32_t index)
	{
		return 1ULL << index;
	}

	std::uint32_t bitCount(std::uint64_t bits)
	{
		return std::bitset<64>(bits).count();
	}

	std::uint64_t lowestBit(std::uint64_t bits)
	{
		return bits & (~bits + 1);
	}

	std::uint64_t highestBit(std::uint64_t bits)
	{
		bits |= bits >>  1;
		bits |= bits >>  2;
		bits |= bits >>  4;
		bits |= bits >>  8;
		bits |= bits >> 16;
		bits |= bits >> 32;
		return bits - (bits >> 1);
	}
}

File: python/bits.py
def to_mask(index):
	return to_flag(index) - 1

def to_flag(index):
	return 1 << index
	
def lowest_bit(i):
	return i & -i

def highest_bit_index(i):
	return i.bit_length() - 1
	
def highest_bit(i):
	return to_flag(highest_bit_index(i))

def bit_count(i):
    return bin(i).count("1")
	
def sparse_bit_count(i):
	count = 0
	while(i):
		i &= i - 1
		count += 1
	return count
